### PR TITLE
test: Code Coverage: Switch Org Component

### PR DESCRIPTION
**Pull request by jayfyle**
### Clickup: https://app.clickup.com/t/85zt04zpk

# Coverage at 96%
switch_org_96

## Notes
1. Some lines for performance tracking remain to be covered and in `ionViewWillEnter` for `navigate_back` and `fromEvent`
2. Certain branches are not getting covered due to edge cases or possibly conditions that may not actually happen in proceed and handle pop up
3. Rest of the branches have been covered by tweaking the current tests

### CODE REVIEW

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 1]
```diff
@@ -1,26 +1,720 @@
...
+    loaderService.showLoader.and.returnValue(Promise.resolve());
+    component.searchOrgsInput.nativeElement.value = 'Staging Loaded';
+    component.searchOrgsInput.nativeElement.dispatchEvent(new Event('keyup'));
+
+    component.ionViewWillEnter();
+    component.filteredOrgs$.subscribe((res) => {
+      expect(res).toEqual(orgData1);
+    });
+    expect(orgService.getOrgs).toHaveBeenCalledTimes(1);
+    expect(orgService.getCurrentOrg).toHaveBeenCalledTimes(1);
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:256]: same for getPrimaryOrg

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 2]
```diff
@@ -1,26 +1,720 @@
...
+    orgService.getOrgs.and.returnValue(of(orgData1));
+    spyOn(component, 'getOrgsWhichContainSearchText').and.returnValue(orgData1);
+    orgService.getCurrentOrg.and.returnValue(of(orgData1[0]));
+    orgService.getPrimaryOrg.and.returnValue(of(orgData2[1]));
+    loaderService.showLoader.and.returnValue(Promise.resolve());
+    component.searchOrgsInput.nativeElement.value = 'Staging Loaded';
+    component.searchOrgsInput.nativeElement.dispatchEvent(new Event('keyup'));
+
+    component.ionViewWillEnter();
+    component.filteredOrgs$.subscribe((res) => {
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:252]: No testcase for choose true/false and single org users

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 3]
```diff
@@ -1,26 +1,720 @@
...
+      })
+    );
+    popoverController.create.and.returnValue(Promise.resolve(popoverSpy));
+    authService.getEou.and.returnValue(Promise.resolve(apiEouRes));
+    component.orgs$ = of(orgData1);
+    fixture.detectChanges();
+
+    await component.showEmailNotVerifiedAlert();
+    expect(authService.getEou).toHaveBeenCalledTimes(1);
+    expect(popoverController.create).toHaveBeenCalledOnceWith({
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:352]: handleDismissPopup

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 4]
```diff
@@ -1,26 +1,720 @@
...
+      expect(router.navigate).toHaveBeenCalledOnceWith(['/', 'post_verification', 'invited_user']);
+    });
+  });
+
+  it('markUserActive(): should mark the user as active and return the org', async () => {
+    loaderService.showLoader.and.returnValue(Promise.resolve());
+    loaderService.hideLoader.and.returnValue(Promise.resolve());
+    orgUserService.markActive.and.returnValue(of(apiEouRes));
+
+    await component
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:390]: await is not needed here.

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 5]
```diff
@@ -1,26 +1,720 @@
...
+      .markUserActive()
+      .pipe(
+        finalize(() => {
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:393]: These assertions should go inside subscribe

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 6]
```diff
@@ -1,26 +1,720 @@
...
+    it('should catch error and clear all caches', async () => {
+      authService.getEou.and.returnValue(Promise.resolve(apiEouRes));
+      loaderService.showLoader.and.returnValue(Promise.resolve());
+      orgService.switchOrg.and.returnValue(throwError(() => {}));
+      secureStorageService.clearAll.and.returnValue(Promise.resolve({ value: true }));
+      storageService.clearAll.and.returnValue(Promise.resolve());
+      loaderService.hideLoader.and.returnValue(Promise.resolve());
+      userEventService.logout.and.returnValue(null);
+
+      await component.switchOrg(orgData1[0]).then(async () => {
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:583]: await and then both are not needed

Comment (yksolanki9) [reply src/app/auth/switch-org/switch-org.page.spec.ts:583]: remove then

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 7]
```diff
@@ -1,26 +1,720 @@
...
+        await Promise.resolve();
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:584]: Why is this needed?

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 8]
```diff
@@ -1,26 +1,720 @@
...
+        expect(loaderService.showLoader).toHaveBeenCalledOnceWith('Please wait...', 2000);
+        expect(authService.getEou).toHaveBeenCalledTimes(1);
+        expect(orgService.switchOrg).toHaveBeenCalledOnceWith(orgData1[0].id);
+        expect(secureStorageService.clearAll).toHaveBeenCalledTimes(1);
+        expect(storageService.clearAll).toHaveBeenCalledTimes(1);
+      });
+
+      expect(loaderService.hideLoader).toHaveBeenCalledTimes(1);
+      expect(userEventService.logout).toHaveBeenCalledTimes(1);
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:593]: Spy and check globalCacheBusterNotifier

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 9]
```diff
@@ -1,26 +1,720 @@
...
+      spyOn(component, 'proceed').and.returnValue(Promise.resolve(null));
+
+      await component.switchOrg(orgData1[0]);
+      expect(loaderService.showLoader).toHaveBeenCalledOnceWith('Please wait...', 2000);
+      expect(authService.getEou).toHaveBeenCalledTimes(1);
+      expect(recentLocalStorageItemsService.clearRecentLocalStorageCache).toHaveBeenCalledTimes(1);
+      expect(component.proceed).toHaveBeenCalledTimes(1);
+      expect(userEventService.clearTaskCache).toHaveBeenCalledTimes(1);
+      expect(component.trackSwitchOrg).toHaveBeenCalledOnceWith(orgData1[0], apiEouRes);
+      expect(orgService.switchOrg).toHaveBeenCalledOnceWith(orgData1[0].id);
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:610]: globalCacheBusterNotifier

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 10]
```diff
@@ -1,26 +1,720 @@
...
+      await component.signOut();
+      expect(authService.getEou).toHaveBeenCalledTimes(1);
+      expect(deviceService.getDeviceInfo).toHaveBeenCalledTimes(1);
+      expect(authService.logout).toHaveBeenCalledOnceWith({
+        device_id: extendedDeviceInfoMockData.uuid,
+        user_id: apiEouRes.us.id,
+      });
+      expect(secureStorageService.clearAll).toHaveBeenCalledTimes(1);
+      expect(storageService.clearAll).toHaveBeenCalledTimes(1);
+      expect(userEventService.logout).toHaveBeenCalledTimes(1);
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:629]: globalCacheBusterNotifier

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 11]
```diff
@@ -1,26 +1,720 @@
...
+    });
+
+    it('should clear cache if sign out fails', async () => {
+      deviceService.getDeviceInfo.and.returnValue(of(extendedDeviceInfoMockData));
+      authService.getEou.and.throwError('Error');
+
+      await component.signOut();
+      expect(secureStorageService.clearAll).toHaveBeenCalledTimes(1);
+      expect(storageService.clearAll).toHaveBeenCalledTimes(1);
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:638]: globalCacheBusterNotifier

**Review by yksolanki9 — CHANGES_REQUESTED** [review]
Are we planning to achieve 100% in a separate PR? Please mention what is covered and what is left in the PR description

File: src/app/auth/switch-org/switch-org.page.spec.ts (hunk 6 above)
Comment (jayfyle) [reply src/app/auth/switch-org/switch-org.page.spec.ts:583]: The method returns a Promise, and multiple nested async blocks hence removing anything will cause the test to fail.

File: src/app/auth/switch-org/switch-org.page.spec.ts [hunk 12]
```diff
@@ -239,24 +240,60 @@ describe('SwitchOrgPage', () => {
...
+    it('should show orgs and setup search bar', async () => {
+      orgService.getOrgs.and.returnValue(of(orgData1));
+      spyOn(component, 'getOrgsWhichContainSearchText').and.returnValue(orgData1);
+      spyOn(component, 'proceed').and.returnValue(Promise.resolve());
+      orgService.getCurrentOrg.and.returnValue(of(orgData1[0]));
+      orgService.getPrimaryOrg.and.returnValue(of(orgData2[1]));
+      loaderService.showLoader.and.returnValue(Promise.resolve());
+      fixture.detectChanges();
+
+      await component.ionViewWillEnter();
```
Comment (yksolanki9) [inline src/app/auth/switch-org/switch-org.page.spec.ts:253]: Why async?

**Review by yksolanki9 — CHANGES_REQUESTED** [review]
Discussed on slack

**Review by yksolanki9 — APPROVED** [review]
